doc/man7/provider-keyexch.pod

```diff
@@ -0,0 +1,180 @@
+exchange algorithms and make them available to applications via the API
+functions L<EVP_PKEY_derive_init_ex(3)>, and L<EVP_PKEY_derive(3)> (as well as
+other related functions).
+
+All "functions" mentioned here are passed as function pointers between
+F<libcrypto> and the provider in B<OSSL_DISPATCH> arrays via
+B<OSSL_ALGORITHM> arrays that are returned by the provider's
+provider_query_operation() function
+(see L<provider-base(7)/Provider Functions>).
+
+All these "functions" have a corresponding function type definition
+named B<OSSL_{name}_fn>, and a helper function to retrieve the
+function pointer from an B<OSSL_DISPATCH> element named
+B<OSSL_get_{name}>.
+For example, the "function" OP_keyexch_newctx() has these:
+
+ typedef void *(OSSL_OP_keyexch_newctx_fn)(void *provctx);
+ static ossl_inline OSSL_OP_keyexch_newctx_fn
+     OSSL_get_OP_keyexch_newctx(const OSSL_DISPATCH *opf);
+
+B<OSSL_DISPATCH> arrays are indexed by numbers that are provided as
+macros in L<openssl-core_numbers.h(7)>, as follows:
+
+ OP_keyexch_newctx           OSSL_FUNC_KEYEXCH_NEWCTX
+ OP_keyexch_freectx          OSSL_FUNC_KEYEXCH_FREECTX
+ OP_keyexch_dupctx           OSSL_FUNC_KEYEXCH_DUPCTX
+
+ OP_keyexch_init             OSSL_FUNC_KEYEXCH_INIT
+ OP_keyexch_set_peer         OSSL_FUNC_KEYEXCH_SET_PEER
+ OP_keyexch_derive           OSSL_FUNC_KEYEXCH_DERIVE
+
+ OP_keyexch_set_params       OSSL_FUNC_KEYEXCH_SET_PARAMS
+
+A key exchange algorithm implementation may not implement all of these functions.
+In order to be a consistent set of functions a provider must implement
+OP_keyexch_newctx, OP_keyexch_freectx, OP_keyexch_init and OP_keyexch_derive.
+All other functions are optional.
+
+A key exchange algorithm must also implement some mechanism for generating,
+loading or importing keys via the key management (OSSL_OP_KEYMGMT) operation.
+See L<provider-keymgmt(7)> for further details.
+
+=head2 Context Management Functions
+
+OP_keyexch_newctx() should create and return a pointer to a provider side
+structure for holding context information during a key exchange operation.
+A pointer to this context will be passed back in a number of the other key
+exchange operation function calls.
+The paramater B<provctx> is the provider context generated during provider
+initialisation (see L<provider(3)>).
+
+OP_keyexch_freectx() is passed a pointer to the provider side key exchange
+context in the B<ctx> parameter.
+This function should free any resources associated with that context.
+
+OP_keyexch_dupctx() should duplicate the provider side key exchange context in
+the B<ctx> parameter and return the duplicate copy.
+
+=head2 Shared Secret Derivation Functions
+
+OP_keyexch_init() initialises a key exchange operation given a newly created
+provider side key exchange context in the B<ctx> paramter, and a pointer to a
+provider key object in the B<provkey> parameter.
+The key object will have been previously generated, loaded or imported into the
```

Comment:
s/will/should/ ?

Here and below.

Again, I'm not fussed if you don't want to change it.

Reply:
I changed it as requested.